---
title: "Multi Tumor Modelling"
output:
  pdf_document: default
  html_notebook:
    code_folding: none
  html_document:
    df_print: paged
---

# Preamble

```{r}
library(plyr)
library(tidyverse)
library(ggplot2)
library(leaps)
library(reshape2)
library(aricode)
library(corrplot)
library(MuMIn)
library(biglm)
library(clusterSim)
source("graphs.R")
```

Load the preprocessed data.

```{r}
load(file = "F:/Data/University/Thesis/Data/CHASMplus/preprocessed_CHASM_data.RData")
projects.names <- names(projects)
```

# Descriptive Analysis

> What is the shape of the data?

## Basic Summary Statistics

Basic statistics (`summary` is a bit messy as it considers all variables):

```{r}
# Sometimes, there are no males, so the function would error.
# This next function prevents that
extract_safe <- function(t, what, default = 0) {
  tryCatch(t[[what]], error = function(e) default)
}

sum_clin <- function(data, name){
  totn <- length(data$ID)
  n0mut <- length(data$ID[data$driver.freq == 0])
  perc0mut <- n0mut/totn*100
  sex <- table(data$gender)/totn*100
  return(
    data.frame(
      # tumor metadata
      ID = name,
      # Numerosities
      numerosity = totn,
      numerosity.0.drivers = n0mut,
      percent.0.drivers = n0mut/totn*100,
      # Driver distributions
      mean.driver.nr = mean(data$driver.freq),
      median.driver.nr = median(data$driver.freq),
      sd.driver.nr = sd(data$driver.freq),
      driver.var.coff = sd(data$driver.freq) / mean(data$driver.freq),
      # Passenger distributions
      mean.passenger.nr = mean(data$passenger.freq),
      median.passenger.nr = median(data$passenger.freq),
      sd.passenger.nr = sd(data$passenger.freq),
      passenger.var.coff = sd(data$passenger.freq) / mean(data$passenger.freq),
      # Sex
      percent.reported.male = signif(extract_safe(sex, "male"), digits=4),
      # Age
      mean.age.years = mean((-data$days_to_birth)/365.25, na.rm = TRUE),
      sd.age.years = sd((-data$days_to_birth)/365.25, na.rm = TRUE)
    )
  )
}

print_clin <- function(clin.stats){
  l <- function(x){return(signif(x, digits = 4))}
  cat(paste(
    "Statistics for clinical data - ", clin.stats$ID, "\n", "\n",
    "Numerosities:\n",
    "\tNumber of patients :          ", clin.stats$numerosity, "\n",
    "\tNr. of pat. with no drivers : ", clin.stats$numerosity.0.drivers, "\n",
    "\t% of pat. with no drivers :   ", l(clin.stats$percent.0.drivers), " %\n",
    "Driver Distribution:\n",
    "\tMean Driver frequency :       ", l(clin.stats$mean.driver.nr), "\n",
    "\tStandard Deviation of Mean :  ", l(clin.stats$sd.driver.nr), "\n",
    "\tVariation Coefficient:        ", l(clin.stats$driver.var.coff), "\n",
    "Passenger Distribution:\n",
    "\tMean passenger frequency :    ", l(clin.stats$mean.passenger.nr), "\n",
    "\tStandard Deviation of Mean :  ", l(clin.stats$sd.passenger.nr), "\n",
    "\tVariation Coefficient:        ", l(clin.stats$passenger.var.coff), "\n",
    "\tMedian passenger frequency:   ", l(clin.stats$median.passenger.nr), "\n",
    "\n",
    "% reported male:  ", clin.stats$percent.reported.male, " %\n",
    "Mean Age (years): ", l(clin.stats$mean.age.years), "\n",
    "SD age (years):   ", l(clin.stats$sd.age.years),
    sep=""
  ))
}

descriptive.stats <- list()
for (name in projects.names) {
  descriptive.stats[[name]] <- sum_clin(projects[[name]]$clinical, name)
}
fused.stats <- descriptive.stats %>% reduce(rbind)
fused.stats
# Print out only GBM long stats
print_clin(descriptive.stats$`TCGA-GBM`)
```
We can now see some distributions.

## Frequency Histograms

```{r, fig.width=6, fig.height=4}
MutationHist <- function(data, title="Number of mutations per patient"){
  library(ggplot2)
  tabulated_freqs <- table(data$driver.freq)
  set_vjust_col <- function(count){
    if(count > 15){
      return("white")
    }else{
      return("blue")
    }
  }
  plot <- ggplot(data, aes(x=driver.freq)) +
    theme_minimal() +
    geom_bar(stat="count", show.legend=FALSE, fill="blue", width=1) +
    scale_x_continuous(breaks=seq(from=0, to=max(data$driver.freq), by=1), minor_breaks = NULL) +
    xlab("Number of Driver Mutations") + ylab("Number of Patients") +
    ggtitle(title)

  pdata <- ggplot_build(plot)
  set_vjust_val <- function(x, t=10){if(x > t){return(2)} else {return(-2)}}
  set_vjust_col <- function(x, t=10){if(x > t){return("white")} else {return("darkblue")}}
  unnested <- pdata[[1]][[1]]
  vjust_values <- unlist(lapply(unnested$count, set_vjust_val, t=max(unnested$count)/10))
  vjust_colours <- unlist(lapply(unnested$count, set_vjust_col, t=max(unnested$count)/10))

  plot <- plot + geom_text(stat='count', aes(label=..count..), vjust=vjust_values, col=vjust_colours)
  print(plot)
}

for (name in projects.names) {
  MutationHist(projects[[name]]$clinical, title = paste("Number of mutations per patient -", name))
}
```

## Passenger - Driver distributions

Now, boxplots of the distributions of passengers by the amount of drivers.

```{r, fig.width=6, fig.height=4}
driverboxplots <- function(clinical.data, project.id, colour = "lightblue"){
  clinical.data$driver.freq <- as.factor(clinical.data$driver.freq)
  plot <- ggplot(clinical.data, aes(group = driver.freq, x = driver.freq, y = passenger.freq)) +
    stat_boxplot(geom ='errorbar', width = 0.4) +
    geom_point(alpha=0) +
    geom_boxplot(fill = colour) +
    scale_y_log10() +
    ylab("Passenger frequency") + xlab("Driver Frequency") +
    ggtitle(paste("Pass/Driver distributions -", project.id)) +
    theme_bw() + theme(panel.grid.minor = element_blank()) +
    annotation_logticks(sides = "l")

  clinical.data$driver.freq <- as.numeric(clinical.data$driver.freq)
  x.box <- axis_canvas(plot, axis = "x", coord_flip = TRUE) +
    stat_boxplot(data = clinical.data, aes(y = driver.freq, x = 1), geom ='errorbar') +
    geom_boxplot(data = clinical.data, aes(y = driver.freq, x = 1), fill = colour) +
    coord_flip()
  suppressMessages(
    y.box <- axis_canvas(plot, axis = "y") +
      # Re-adding the log10 scale generates a warning, so I'm suppressing it
      scale_y_log10() +
      stat_boxplot(data = clinical.data, aes(y = passenger.freq, x = 1), geom ='errorbar') +
      geom_boxplot(data = clinical.data, aes(y = passenger.freq, x = 1), fill = colour)
  )

  empty <- ggdraw()

  final.plot <- insert_xaxis_grob(plot, x.box, grid::unit(7, "mm"), position = "top")
  final.plot <- insert_xaxis_grob(final.plot, empty, grid::unit(2, "mm"), position = "top")

  final.plot <- insert_yaxis_grob(final.plot, y.box, grid::unit(7, "mm"), position = "right")
  final.plot <- insert_yaxis_grob(final.plot, empty, grid::unit(2, "mm"), position = "right")

  return(ggdraw(final.plot))
}

for (name in projects.names) {
  print(driverboxplots(projects[[name]]$clinical, name, colour = "lightgreen"))
}
```
Margin boxplots are computed on the overall distribution. TCGA-ACC values show some problems, and I'm not sure why. Probably the driver margin boxplot.

#

# Models

> Are the number of drivers and passenger for each patient modelled well by the clinical variables?

## Variable availability

The various variables are not all complete. The models need complete cases, so the variables that are missing in many of the cases need to be removed.

First I want to look at the availabilities for the clinical data, that is, the amount of non-NA values in each of the clinical data variables:

```{r}
get_availability <- function(x, cutoff, varname="frac.availability", cut_0_mut = FALSE){
  #' Produce a dataframe containing the relative percent availability of clinical
  #' data (i.e. the % reported fields)
  #'
  #' @param x Data as created by project.py clinical frequency function
  #' @param cutoff Remove all fields with % less than this value.
  #' @param cut_0_mut If true, removes patients with 0 driver mutations.
  #' @return Data availability frame
  if(cut_0_mut){
    x <- subset(x, x$frequency != 0)
  }
  # Numbers holds the amount of non-na values in the dataframe
  numbers <- colSums(!is.na(x))
  numbers <- numbers/length(x[,1])*100
  numbers <- subset(numbers, numbers >= cutoff)
  frame <- data.frame(numbers)
  colnames(frame) <- varname
  return(frame)
}

for (name in projects.names) {
  projects[[name]]$availability <- get_availability(projects[[name]]$clinical, cutoff=0, varname = name)
}
```

To visualize them better, I'll make them in a single dataframe.

```{r, include=FALSE}
availabilities <- list()
for (name in projects.names) {
  availabilities[[name]] <- rownames_to_column(projects[[name]]$availability, var="ID")
}
fused.avail <- reduce(availabilities, merge, by = "ID", all = TRUE)
fused.avail
```

To show only those variables available for each tumor type, I filter the above frame to keep only those variables that are available (>90%) in all tumours.

```{r}
modelling.data <- projects

for (name in projects.names) {
  modelling.data[[name]]$avail.variables <- rownames(subset(projects[[name]]$availability, projects[[name]]$availability > 90))
}
```

The clinical variables are detailed below. Source for this data is the GDC data dictionary. I use the Demographic, Diagnosis and Exposures clinical data. There are other types of data in the portal (Family History, Follow ups and more), but they are one-to-many (for instance, one patient might have more than one family member with cancer, or more than one follow up), so they have to be aggregated in some way.

## Variable Description

The variables present in the data, in at least one tumor type after filtering, are as follows:

```{r}
vars.lst <- list()
for (name in projects.names){
  vars.lst[[name]] <- modelling.data[[name]]$avail.variables
}

all.vars <- sort(unique(reduce(vars.lst, c)))
all.vars
```

Ignoring the data columns `driver.freq`, `passenger.freq` and `ID`, we have a total number of 39 clinical variables that are available in at least one tumor type. These are:

* **Demographic**
  * `age_at_index` : (int) The patient's age (in years) on the reference or anchor date date used during date obfuscation.
  * `days_to_birth` : (int) Number of days between the date used for index and the date from a person's date of birth represented as a calculated negative number of days.
  * `ethnicity` : (Enum) [__required__] An individual's self-described social and cultural grouping, specifically whether an individual describes themselves as Hispanic or Latino. The provided values are based on the categories defined by the U.S. Office of Management and Business and used by the U.S. Census Bureau.
  * `gender` : (Enum) [__required__] Text designations that identify gender. Gender is described as the assemblage of properties that distinguish people on the basis of their societal roles. Self-reported.
  * `race` : (Enum) [__required__] An arbitrary classification of a taxonomic group that is a division of a species. It usually arises as a consequence of geographical isolation within a species and is characterized by shared heredity, physical attributes and behavior, and in the case of humans, by common history, nationality, or geographic distribution. The provided values are based on the categories defined by the U.S. Office of Management and Business and used by the U.S. Census Bureau.
  * `vital_status` : (Enum) [__required__] The survival state of the person registered on the protocol.
  * `year_of_birth` : (int) Numeric value to represent the calendar year in which an individual was born.
* **Diagnosis**:
  * `age_at_diagnosis` : (int) [__required__] Age at the time of diagnosis expressed in number of days since birth.
  * `ajcc_clinical_m` : (Enum) Extent of the distant metastasis for the cancer based on evidence obtained from clinical assessment parameters determined prior to treatment.
    * MX: Cannot be measured
    * M0: Cancer has not spread to other parts of the body
    * cM0 (i+): Minimal cancer spread (less than 0.2mm areas of cancer spread)
    * M1/M1a/M1b/M1c: Cancer has spread to other parts of the body
  * `ajcc_clinical_n` : (Enum) Extent of the regional lymph node involvement for the cancer based on evidence obtained from clinical assessment parameters determined prior to treatment.
    * NX: Cannot be measured
    * N0: There is no cancer in nearby lymph nodes
    * N1-4: Number and location of lymph nodes that show the presence of cancer cells. Several variations exist, such as N0 (1+/-), N11/b/c, etc...
  * `ajcc_clinical_stage` : (Enum) Stage group determined from clinical information on the tumor (T), regional node (N) and metastases (M) and by grouping cases with similar prognosis for cancer. Stage I to IV, with several possible modifiers
  * `ajcc_clinical_t` : (Enum) Extent of the primary cancer based on evidence obtained from clinical assessment parameters determined prior to treatment.
    * TX: Cannot be measured
    * T0: Cannot be found
    * T1-4: Larger stages mean large cell mass and/or penetration into tissues
  * `ajcc_pathologic_m` : (Enum) Code to represent the defined absence or presence of distant spread or metastases (M) to locations via vascular channels or lymphatics beyond the regional lymph nodes, using criteria established by the American Joint Committee on Cancer (AJCC). Same levels as `ajcc_clinical_m`.
  * `ajcc_pathologic_n` : (Enum) The codes that represent the stage of cancer based on the nodes present (N stage) according to criteria based on multiple editions of the AJCC's Cancer Staging Manual. Same levels as `ajcc_clinical_n`.
  * `ajcc_pathologic_stage` : (Enum) The extent of a cancer, especially whether the disease has spread from the original site to other parts of the body based on AJCC staging criteria. Same levels as `ajcc_clinical_stage`.
  * `ajcc_pathologic_t` : () Code of pathological T (primary tumor) to define the size or contiguous extension of the primary tumor (T), using staging criteria from the American Joint Committee on Cancer (AJCC). Same levels as `ajcc_clinical_t`.
  * `ajcc_staging_system_edition` : (Enum) The text term used to describe the version or edition of the American Joint Committee on Cancer Staging Handbooks, a publication by the group formed for the purpose of developing a system of staging for cancer that is acceptable to the American medical profession and is compatible with other accepted classifications. There are 8 possible editions, from 1st to 8th.
  * `ann_arbor_b_symptoms` : (Binary Enum) Text term to signify whether lymphoma B-symptoms are present as noted in the patient's medical record.
  * `ann_arbor_extranodal_involvement` : (Enum) The text term used to describe the clinical classification of lymphoma, as defined by the Ann Arbor Lymphoma Staging System.
  * `days_to_diagnosis`: (int) Number of days between the date used for index and the date the patient was diagnosed with the malignant disease.
  * `days_to_last_follow_up` : (int) Time interval from the date of last follow up to the date of initial pathologic diagnosis, represented as a calculated number of days.
  * `figo_stage` : (Enum) The extent of a cervical or endometrial cancer within the body, especially whether the disease has spread from the original site to other parts of the body, as described by the International Federation of Gynecology and Obstetrics (FIGO) stages.
  * `icd_10_code`: (str) Alphanumeric value used to describe the disease code from the tenth version of the International Classification of Disease (ICD-10).
  * `masaoka_stage` : (Enum) The text term used to describe the Masaoka staging system, a classification that defines prognostic indicators for thymic malignancies and predicts tumor recurrence.
  * `morphology` : (Enum) [__required__] The third edition of the International Classification of Diseases for Oncology, published in 2000 used principally in tumor and cancer registries for coding the site (topography) and the histology (morphology) of neoplasms. The study of the structure of the cells and their arrangement to constitute tissues and, finally, the association among these to form organs. In pathology, the microscopic process of identifying normal and abnormal morphologic characteristics in tissues, by employing various cytochemical and immunocytochemical stains. A system of numbered categories for representation of data.
  * `primary_diagnosis` : (Enum) [__required__] Text term used to describe the patient's histologic diagnosis, as described by the World Health Organization's (WHO) International Classification of Diseases for Oncology (ICD-O).
  * `primary_gleason_grade` : (Enum) The text term used to describe the primary Gleason score, which describes the pattern of cells making up the largest area of the tumor. The primary and secondary Gleason pattern grades are combined to determine the patient's Gleason grade group, which is used to determine the aggresiveness of prostate cancer. Note that this grade describes the entire prostatectomy specimen and is not specific to the sample used for sequencing.
  * `prior_malignancy` : (Enum)  The yes/no/unknown indicator used to describe the patient's history of prior cancer diagnosis.
  * `prior_treatment` : (Enum) A yes/no/unknown/not applicable indicator related to the administration of therapeutic agents received before the body specimen was collected.
  * `site_of_resection_or_biopsy` : (Enum) [__required__] The text term used to describe the anatomic site of origin, of the patient's malignant disease, as described by the World Health Organization's (WHO) International Classification of Diseases for Oncology (ICD-O).
  * `secondary_gleason_grade` : (Enum) The text term used to describe the secondary Gleason score, which describes the pattern of cells making up the second largest area of the tumor. The primary and secondary Gleason pattern grades are combined to determine the patient's Gleason grade group, which is used to determine the aggresiveness of prostate cancer. Note that this grade describes the entire prostatectomy specimen and is not specific to the sample used for sequencing.
  * `synchronous_malignancy` : (Binary Enum) A yes/no/unknown indicator used to describe whether the patient had an additional malignant diagnosis at the same time the tumor used for sequencing was diagnosed. If both tumors were sequenced, both tumors would have synchronous malignancies.
  * `tissue_or_organ_of_origin` : (Enum) [__required__] The text term used to describe the anatomic site of origin, of the patient's malignant disease, as described by the World Health Organization's (WHO) International Classification of Diseases for Oncology (ICD-O).
  * `year_of_diagnosis` : (int) Numeric value to represent the year of an individual's initial pathologic diagnosis of cancer.
* **Exposures**:
  * `alcohol_history` : (Binary Enum) A response to a question that asks whether the participant has consumed at least 12 drinks of any kind of alcoholic beverage in their lifetime.
* **Follow Up**:
  * `bmi` : (number) A calculated numerical quantity that represents an individual's weight to height ratio.
  * `height` : (number) The height of the patient in centimeters.
  * `weight` : (number) The weight of the patient measured in kilograms.

Note that I cannot find the `tumor_stage` variable in the GDC data portal, however it looks like, from cursory inspection, that it is identical to the `ajcc_pathologic_stage`. I'm not sure where it comes from...

### Binary mutation Presence

Martina mixed gene level and mutation level in her models, but I want to keep them separate. I'd need some reshaping to make the patients have a string of binary values to use as labels.

```{r}
reshape_crystal <- function(project){
  relevant <- project$crystal[,c("ID", "Renamed.HUGO", "Hugo")]
  # I stole this from stack overflow
  gene_labels <- relevant %>%
    pivot_longer(ends_with("hugo"), names_to = "genes") %>%
    complete(ID, value) %>%
    dplyr::mutate(genes = as.numeric(!is.na(genes))) %>%
    pivot_wider(names_from = value, values_from = genes, values_fn = max)
  # Putting `all.x=TRUE, all.y=TRUE` makes some NAs appear, which is weird....
  # TODO: Check if the length of the gene labels matrix is the same as the crystal,
  # but nothing should have been lost by the trasposition
  gene_labels <- merge(gene_labels, project$clinical[,c("ID", "driver.freq", "passenger.freq")], all = TRUE)
  # See disclaimer above. I think that the NAs come from 0-driver patients, that have no ID in the
  # pivoted gene_labels table, so the real labels should be zero
  gene_labels[is.na(gene_labels)] <- 0
  return(gene_labels)
}

for (name in names(projects)) {
  # I set an arbitrary cutoff of 0.01% in order to remove the cols with 0 availability
  projects[[name]]$gene_labels <- reshape_crystal(projects[[name]])
}
```

Now we have available as variables the presence or absence of each mutations present in the crystal.

## Clustering

My question is if the data clusters based on the types of mutations present in the tumours. The first thing to consider is that if the mutation occurs in just a single patient, then it doesn't form a cluster. I set a threshold of being present in at least 10% of all patients to be considered, and 1 at minimum.

There is the need to have some method of cluster quality analysis, so we use the DB index.

```{r, fig.width=6, fig.height=4}
get.cluster.distances <- function(project){
  possible.levels <- project$gene_labels
  possible.levels <- possible.levels[!colnames(possible.levels) %in% c("ID", "driver.freq", "passenger.freq")]
  possible.levels <- possible.levels[colSums(possible.levels) > max(length(possible.levels[1]) / 10, 25)]
  possible.levels.name <- colnames(possible.levels)
  if (length(possible.levels.name) == 0) {
    return(NULL)
  }
  # This is needed as the function expects clusters with names 1, 2, 3... not 0, 1, 2...
  possible.levels <- possible.levels + 1
  data <- project$gene_labels[, c("passenger.freq", "driver.freq")]

  distances <- list()

  for (name in possible.levels.name) {
    distances[[name]] <- index.DB(data, possible.levels[[name]])$DB
  }
  vector.dist <- reduce(distances, c)
  names(vector.dist) <- names(distances)
  return(vector.dist)
}

plot_distance <- function(labels, variable, name, DB_value){
  # I set the seed here so the different plots can be compared as they
  # will have the same jitter displacement
  set.seed(72)

  p <- ggplot(labels, aes(passenger.freq, driver.freq)) +
    geom_jitter(alpha=0.7, height=0.2, aes(colour = as.factor(.data[[variable]]))) +
    scale_x_log10() +
    ylim(c(-0.5, max(labels$driver.freq + 0.5))) +
    ggtitle(paste("Clustering by mutation", variable, name, paste("DB:", DB_value), sep = " - ")) +
    xlab("Passenger Frequency") + ylab("Driver Frequency") +
    annotation_logticks(sides = "b") +
    labs(color = "Mut?") +
    theme_bw() + theme(panel.grid.minor = element_blank())

  # Need a new frame with just the driver levels, and the associated non-coloured and coloured sums.
  #labels %>% dplyr::select(!c(passenger.freq, ID)) %>% melt(id.vars = c("driver.freq")) -> bdata
  #bdata %>% group_by(driver.freq) %>% summarise(count = n()) -> box.data

  labels %>% dplyr::select(!c(passenger.freq, ID)) %>% melt(id.vars = c("driver.freq")) -> bdata
  var <- variable
  bdata %>% group_by(driver.freq, value) %>% subset(variable == var) %>% summarise(count = n()) -> box.data

  y.box <- axis_canvas(p, axis = "y", box.data, aes(fill = as.factor(value), x = driver.freq, y = count), coord_flip = TRUE) +
    geom_bar(position = "fill", stat = "identity", width = 0.8) +
    coord_flip()

  final.plot <- insert_yaxis_grob(p, y.box, grid::unit(15, "mm"), position = "right")
  final.plot <- insert_yaxis_grob(final.plot, ggdraw(), grid::unit(2, "mm"), position = "right")

  ggdraw(final.plot)
}

make_cluster_plots <- function(project, name="", max_plots = 5, crop_dist = 90) {
  # Crops the data distribution at the 90th percentile to remove outliers
  project$gene_labels <- subset(project$gene_labels, project$gene_labels$passenger.freq < quantile(project$gene_labels$passenger.freq, crop_dist/100))
  dists <- get.cluster.distances(project)
  if (is.null(dists)) {
    stop(paste("No viable distances to plot for", name))
  }
  dists <- sort(dists, decreasing = FALSE)

  names(dists) %>% head(max_plots) -> to_plot
  for (level in to_plot) {
    print(plot_distance(project$gene_labels, level, name, signif(dists[level], 3)))
  }
}

# TCGA-ACC crashes it, for some reason.
for (name in p.names) {
  tryCatch(make_cluster_plots(projects[[name]], name, 2), error = function(e) {print(e)})
}
```

Some plots fail - and I cannot figure out what is causing the error. Some subscript somewhere?

```{r}
make_cluster_plots(projects$`TCGA-PAAD`, "PAAD", 5)
```
